evm/p2p/discovery.py

```diff
@@ -0,0 +1,600 @@
+
+## Joining the network
+
+When joining the network, fills its node table by perfoming a
+recursive Find Node operation with its own ID as the `Target`. The
+initial Find Node request is sent to one or more bootstrap nodes.
+
+## RPC Protocol
+
+RLPx nodes that want to accept incoming connections should listen on
+the same port number for UDP packets (Node Discovery Protocol) and
+TCP connections (RLPx protocol).
+
+All requests time out after are 300ms. Requests are not re-sent.
+
+"""
+
+
+# TODO(gsalgado): draw a dependency diagram of those things and see how it can be improved;
+# currently:
+#    DiscoveryProtocol contains a KademliaProtocol (.kademlia)
+#        -> also has a reference to NodeDiscovery (.transport)
+#        -> all recv_* methods here delegate to KademliaProtocol
+#    KademliaProtocol contains a DiscoveryProtocol (.wire)
+#        -> .ping(), .recv_ping(), find_node(), recv_neighbours(), recv_find_node()
+#           and _query_neighbours() delegate to DiscoveryProtocol
+class DiscoveryProtocol(asyncio.DatagramProtocol):
+
+    """
+    ## Packet Data
+    All packets contain an `Expiration` date to guard against replay attacks.
+    The date should be interpreted as a UNIX timestamp.
+    The receiver should discard any packet whose `Expiration` value is in the past.
+    """
+    bootstrapped = False
+    transport = None
+    version = 4
+    expiration = 60  # let messages expire after N secondes
+    cmd_id_map = dict(ping=1, pong=2, find_node=3, neighbours=4)
+    rev_cmd_id_map = dict((v, k) for k, v in cmd_id_map.items())
+
+    # number of required top-level list elements for each cmd_id.
+    # elements beyond this length are trimmed.
+    cmd_elem_count_map = dict(ping=4, pong=3, find_node=2, neighbours=2)
+
+    encoders = dict(cmd_id=chr,
+                    expiration=rlp.sedes.big_endian_int.serialize)
+
+    decoders = dict(cmd_id=safe_ord,
+                    expiration=rlp.sedes.big_endian_int.deserialize)
+
+    def __init__(self, privkey, address, bootstrap_nodes):
+        self.privkey = privkey
+        self.pubkey = private_key_to_public_key(self.privkey)
+        self.address = address
+        self.bootstrap_nodes = bootstrap_nodes
+        self.this_node = Node(self.pubkey, address)
+        self.nodes = LRUCache(2048)   # nodeid->Node,  fixme should be loaded
+        self.kademlia = kademlia.KademliaProtocol(self.this_node, wire=self)
+        self.nat_upnp = add_portmap(address.udp_port, 'UDP', 'Ethereum DEVP2P Discovery')
+        log.info('starting discovery proto on', this=self.this_node)
+
+    def listen(self, loop):
+        return loop.create_datagram_endpoint(
+            lambda: self, local_addr=(self.address.ip, self.address.udp_port))
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+    @asyncio.coroutine
+    def bootstrap(self):
+        if self.bootstrapped or len(self.bootstrap_nodes) == 0:
```

Comment:
These both feel like anti-patterns.

* The `self.bootstrapped` state is something that I'd like to look for an alternate solution which didn't require storing state (low priority)
* The *silent* return  if the list of bootstrap nodes feels like a silent failure.

I don't have enough knowledge yet to know if this is a dumb solution but...

What do you think about modifying this so that it instead takes an iterable of bootstrap nodes as an argument.  This would remove the silent return when there are no bootstrap nodes and allow moving that check higher in the stack to allow the `DiscoveryProtocol` object to not care about that stuff as well as possibly removing the need to keep the `self.bootstrapped` state around?